### ISSUE TITLE
Error plot brushing

### ISSUE THREAD

**Issue by jaelafield** [herbie-fp/odyssey#108]
Implement brushing on the Error Plot to give users a focused view on a subset of points.

- [x] Research feasibility and design considerations
- [x] Proof of concept / basic implementation to give visual starting point for feedback
- [x] Discuss priorities and plan components of feature
- [x] Figure out translation from d3 selection on plot to points on graph
- [x] Implement!

**Comment by jaelafield** [comment]
Next planned step is filming a brief video with visual examples of the kind of options with the brushing feature and writing up a few initial corresponding questions, then sending those out for feedback in the herbie channel. Goal to do that is morning of 8/28, though haven't done much on this besides previously discussed planning!

**Comment by elmisback** [comment]
Assuming that's 8/29, that feedback plan sounds good to me! I think we can already plan to implement subset selection with what you've got, pending the solution of the bucketing issue. We can discuss in the meeting tomorrow.

**Comment by jaelafield** [comment]
Update: [Thread](https://uw-cse.slack.com/archives/C0G44MH4P/p1725323114471949) with video of initial design considerations and discussion. General plan for details of first implementation summarized there. New related step:

- [ ] Transfer those notes in a design-doc style to existing brushing [doc](https://docs.google.com/document/d/1i9MhAA0LCN_PTbcNMva1XfAnpTLjEi8d-NoObgI3P5A/edit)

**Comment by jaelafield** [comment]
Notes from 9/12 meeting: 
- Grab subset of points selected (Point type not plot circles)
  (but not until brushing is totally complete, utilize small delay?)
- Make new analysis call, passing in subset of points, to get local error plot for average of points

- If single point is selected, have that replace brushed area, and vice versa

UI notes:
- Make brush more transparent

**Comment by jaelafield** [comment]
Current in-progress (besides above):
- [ ] adjusting brushing to make sure it works with multiple expressions
- [ ] actually pushing brushing to main to get more eyes on it --> send note to folks that there have been `package`/`-lock.json` changes
- [ ] fixing removal of brushing (currently only works with dblclick and it's a little weird, single click partially deselects but not all)

Lower priority UI adjustments to keep in mind for later:
- [ ] adding Selected point
- [ ] considering Pavel's feedback:
>  can we make the "default" state, where you haven't brushed, the same as brushing to select the whole region? (Appear the same way in the UI)
> this might mean presenting the input range differently (making it always use two pins at either side) and also maybe add a background to the not selected region instead of the selected one

**Comment by jaelafield** [comment]
Notes about difficulty with updating global variable causing re-render:
- see if there's a "mouse up" or other action to see when brushing ends (while still retaining selection)
- if "mouse down" bail
- if trying to do continuously: consider <input> implementation style of retaining focus
- make global variable to write to when you want to update
- keep svg itself to try to keep from rerendering, just rerender the content of svg
- transient vs permanent state

**Comment by jaelafield** [comment]
"mouse up" action can be detected with `.on("end", () => {...` originally I was using `.on("brush end", () => {...`
"brush end" may be the correct behavior for the highlighting as it's nice for that to be continuous as the mouse drags, but just "end" can be defined separately and may be good for helping with the global variable setting, however that doesn't fix it on its own.

**Comment by jaelafield** [comment]
(I'm sick, so not much progress) Have an in-progress attempt to save the brushing state so the selected subset of points can be stored globally and then the brushing can be reapplied after the re-render. Have yet to try to debug it, so unsure if it'll work.

**Comment by jaelafield** [comment]
Monday, I re-familiarized myself with what I was attempting earlier last week ^^, and I roughly diagnosed the current bugs, but I also reminded myself how ugly my current approach is. I expect to have a solid hour tomorrow before RA meeting to attempt those fixes I have in mind, but if it works that really just feels like a proof of concept to me, I'd love to find a cleaner way to handle it.

(Covid-catch up on all other things has left me with less time on this than I'd like, so sorry!)

**Comment by jaelafield** [comment]
Working on an attempt on the pattern we talked about in 2:1 yesterday.

This portion (the easy part) works fine with a play value for the subset analysis:
```
global context exists for subset of points analysis
...
in the `ExpressionTable` for each expression where the "error" and "speedup" are calculated
     if a subset of points have been selected
               instead calculate values to be rendered using subset analysis instead of standard analysis
```
I need to actually get that analysis value correctly which I will continue to work on before the meeting tomorrow!

Not currently blocked, though I will update here if something comes up as I continue to work on it.

**Comment by jaelafield** [comment]
I have an analysis call working on just the subset of points selected. 
If I have some free time before tomorrow, I'll try to get the "cost" call working as well. I haven't looked into it much, so if it's a similar pattern it should be quick, otherwise I'm unsure.

There's also some bugginess with deselecting of a brushed region, so I'll try to debug that by tomorrow as well!

**Comment by jaelafield** [comment]
Created flag to hide brushing implementation for feature freeze

**Comment by jaelafield** [comment]
Plan by meeting tomorrow: 

- Will make sure all my (working) code is up to date and in main, still hidden behind the flag I made before, for now. I have a local branch and some stashed things, so I just need to get it organized and together again.
- Will think about/work on how brushing will effect Error plots for multiple variables. I may have some questions on that tomorrow.
- Since the analysis call has been successful, I am returning to the actual brushing appearance. I doubt I'll have much of anything new finished to demo by tomorrow though.

**Comment by jaelafield** [comment]
Last Sunday I was pretty sure I had gotten the main functionality done, but then discovered a weird bug with the analysis that I hadn't caught before. I've just merged in my progress anyway. (It doesn't break anything existing, the analysis just doesn't look quite right when you brush something). 

**If there's time (and I know this is last minute), I could benefit from talking about it some in the RA meeting today!**

Besides that, I expect we'll need to do a round of feedback to see if we like what I have so fair, and I am slightly concerned about the performance, so we can discuss that then too.

**Comment by elmisback** [comment]
Regarding the confusion about dataflow you mentioned in the meeting, I think there may be an issue where the backend is resampling or using a different sample in the analysis call. I can't remember this well, though. We should post this concern as an issue and collect evidence in one place for what you're seeing. If we believe it is due to an issue on the server side we can send the issue to someone working on the Herbie API.

**Comment by jaelafield** [comment]
As far as I can tell this probably isn't a Herbie API issue. The bug has to do with sample data that was originally received from Herbie but shouldn't have changed since then. I think I just need to trace the data better, I think I'm assuming things are the same somewhere when they aren't.

**Comment by elmisback** [comment]
If you need to meet to go over the code, let me know and we can set it up

**Comment by elmisback** [comment]
I think the final step here will be addressing the performance issue in https://github.com/herbie-fp/odyssey/issues/229 and then we can close this!

**Comment by elmisback** [comment]
(updates for this issue will appear on #229, this issue is essentially closed otherwise)

**Comment by jaelafield** [comment]
Finished the last of those small bug fixes! There are some lingering styling things that were discussed way back when, but generally considered low priority for now, so I'm closing the issue!
